src/extension/neoExpress/neoExpressInstaller.ts

```diff
@@ -0,0 +1,181 @@
+import * as vscode from "vscode";
+import PackageVersion from "./packageVersion";
+import workspaceFolder from "../util/workspaceFolder";
+import Log from "../util/log";
+import { findPackage, DotNetPackage, VersionMatchCriteria, PackageLocation, locationString } from "./dotnetToolPackage";
+import { installCommand, updateCommand } from "./dotNetToolCommand";
+import { getIncludeBuildServerFeed, getIncludePreviewReleases } from "../util/packageJsonHelper";
+
+export enum UpdateResult {
+  notUpdated,
+  updated,
+  declinedToUpdate,
+  noNewVersionFromNuget,
+  currentPackageIsNewer,
+  declinedToUpdateAndDontAskAgain,
+}
+
+const LOG_PREFIX = "NeoExpress";
+
+export default class NeoExpressInstaller {
+  readonly name: string;
+  readonly version: string;
+  readonly rootFolder: string;
+  readonly targetPackage: DotNetPackage;
+  readonly context: vscode.ExtensionContext;
+  readonly neverShowAgainKey = "neverShowAgain";
+
+  constructor(context: vscode.ExtensionContext, version: string) {
+    this.context = context;
+    this.name = "neo.express";
+    this.version = version;
+    this.targetPackage = {
+      name: this.name,
+      version: PackageVersion.parse(this.version),
+    };
+    this.rootFolder = workspaceFolder() || process.cwd();
+  }
+
+  async run(): Promise<DotNetPackage | null> {
+    let currentPackage = await findPackage(this.targetPackage, VersionMatchCriteria.nameOnly);
+    if (currentPackage) {
+      const updateOutput = await this.tryUpdate(currentPackage, {
+        ...this.targetPackage,
+        location: currentPackage.location,
+      });
+      if (updateOutput.package) {
+        currentPackage = updateOutput.package;
+      }
+    } else {
+      currentPackage = await this.tryInstall();
+    }
+    return currentPackage;
+  }
+
+  async tryInstall(): Promise<DotNetPackage | null> {
+    let installedPackage = null;
+    try {
+      const selectedLocation = await this.selectUserPreferredLocation();
+      if (selectedLocation === null) {
+        return null;
+      }
+      installedPackage = { ...this.targetPackage, location: selectedLocation };
+      await installCommand(this.rootFolder, installedPackage, getIncludeBuildServerFeed());
+      vscode.window.showInformationMessage(
```

Comment:
Maybe we should use `window.withProgress` to show that the installer is working in the background?

Reply:
added